perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/MembersManagerBlImpl.java

```diff
@@ -2523,25 +2525,38 @@ private String createUserAccountInExternalSystem (PerunSession session, User use
 	}
 
 	/**
-	 * Execute process which expires member and removes his sponsorship
+	 * Execute process which expires member and removes his sponsorship.
 	 *
 	 * @param sess perun session
 	 * @param sponsoredMember who will be processed
+	 * @param vo of sponsored member
 	 */
-	private void processMemberAfterRemovingLastSponsor(PerunSession sess, Member sponsoredMember) {
+	private void processMemberAfterRemovingLastSponsor(PerunSession sess, Member sponsoredMember, Vo vo) {
 
-		//Set member's expiration to today and set status to expired.
+		//Check whether the member should be expired according to vo's expiration rules.
+		boolean cannotBeExpired = false;
 		try {
-			Attribute expiration = getPerunBl().getAttributesManagerBl().getAttribute(sess, sponsoredMember, EXPIRATION);
-			expiration.setValue(BeansUtils.getDateFormatterWithoutTime().format(new Date()));
-			getPerunBl().getAttributesManagerBl().setAttribute(sess,sponsoredMember,expiration);
-		} catch (WrongAttributeAssignmentException | AttributeNotExistsException| WrongAttributeValueException | WrongReferenceAttributeValueException e) {
-			throw new InternalErrorException("cannot set expiration date to today for sponsored member "+sponsoredMember.getId(),e);
-		}
-		try {
-			expireMember(sess, sponsoredMember);
-		} catch (WrongReferenceAttributeValueException | WrongAttributeValueException ex) {
-			throw new InternalErrorException("cannot expire member "+sponsoredMember.getId(),ex);
+			Attribute expirationRules = getPerunBl().getAttributesManagerBl().getAttribute(sess, vo, EXPIRATION_RULES_ATTR);
+			Map<String, String> rulesMap = expirationRules.valueAsMap();
+			if (rulesMap != null && rulesMap.get(SPONSORED_MEMBERS_EXPIRATION_KEY) != null)
+				cannotBeExpired = Boolean.parseBoolean(rulesMap.get(SPONSORED_MEMBERS_EXPIRATION_KEY));
+		} catch (AttributeNotExistsException | WrongAttributeAssignmentException e) {
+			throw new InternalErrorException("Cannot determine whether a previously sponsored member " + sponsoredMember.getId() + "should expire.", e);
```

Comment:
I would think about this exception. Maybe if there is a problem, we should log it as error and continue as default = expire members.

Reply:
Done.